### PR TITLE
Fix CI logo URL validation logic

### LINKED ISSUE #37: Fix CI logo URL validation logic
### Expected behavior

The CI when validating the logo URLs must consider the following:

- If the specified logo URL is external to GitHub, perform normal validation
- If the specified logo URL is pointing to GitHub repo
  - It must be in the app-registry repo only
  - If the URL points to the main branch, first check if the images are available on the current PR
  - If not, check on the main branch
  - If it doesn't exist anywhere, throw a validation error
  - If it exists on the PR, if possible, validate the image resolution to match the recommended resolution of `64px x 64px`

### Actual behavior

It blindly validates for the specified URL and fails the CI for first time metadata submissions with URLs pointing to the app-registry GH repo on the main branch

### Steps to reproduce

- https://github.com/LiskHQ/app-registry/actions/runs/5654493429/job/15318913600?pr=33

### Which version(s) does this affect? (Environment, OS, etc...)

NA

### Additional Information

- Update documentation to highlight the following as important information:
  - All the specified deployments must the available: appNodes, Lisk Service URLs
  - The sidechains must be registered on the mainchain before the metadata submission
  - If the logos are pointing to GitHub it must be in the app-registry repo

### PR DESCRIPTION
**Pull request by vardan10**
### What was the problem?
This PR resolves #37 

### How was it solved?
- [x] Added validation for logos incase supplied with PR or already present in the repo
- [x] Updated documentation
- [x] Updated logs to highlight errors properly in ci.

### How was it tested?
- [x] Locally with act

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -24,7 +24,9 @@ All the *app.json* files within the repository must adhere to the schema specifi
 Example: *mainnet/Lisk/app.json*
 
 *Important information*:
-- All the specified URLs in *app.json* (i.e. `genesisURL`, `appNodes`, and `serviceURLs`) are validated in real-time. Hence, please ensure that they are valid and reachable at the time of submission.
+- All the specified URLs in *app.json* (i.e. `genesisURL`, `appNodes`, and `serviceURLs`, etc.) are validated in real-time. Hence, please ensure that they are valid and reachable at the time of submission.
+- Prior to metadata submission, ensure sidechains are registered on the mainchain.
+- For logos pointing to app-registry GitHub URL, use the main branch in the URL to ensure working functionality after PR merge. All logos submitted needs to have `64x64` resolution.
```
Comment (sameersubudhi) [inline README.md:29]:
```suggestion
- Prior to the metadata submission, please ensure that the Lisk applications are registered on the Lisk mainchain and vice-versa.
- For logos pointing to app-registry GitHub URL, use the main branch in the URL to ensure working functionality after PR merge. All logos submitted need to adhere to `64px x 64px` resolution.
```

File: README.md [hunk 2]
```diff
@@ -38,6 +40,10 @@ All the *nativetokens.json* files within the repository must adhere to the schem
 
 Example: *mainnet/Lisk/nativetokens.json*
 
+*Important information*:
+- Prior to metadata submission, ensure sidechains are registered on the mainchain.
+- For logos pointing to app-registry GitHub URL, use the main branch in the URL to ensure working functionality after PR merge. All logos submitted needs to have `64x64` resolution.
```
Comment (sameersubudhi) [inline README.md:45]:
```suggestion
- Prior to the metadata submission, please ensure that the Lisk applications are registered on the Lisk mainchain and vice-versa.
- For logos pointing to app-registry GitHub URL, use the main branch in the URL to ensure working functionality after PR merge. All logos submitted need to adhere to `64px x 64px` resolution.
```

File: validation/src/validateURLs.js [hunk 3]
```diff
@@ -62,29 +64,103 @@ const validateProjectPageURL = async (projectPageURL) => {
...
+const validateImageResolution = async (params) => {
+	const validationErrors = [];
+
+	const { url, filePath } = params;
+	let imageBuffer;
+
+	try {
+		if (url) {
+			const response = await httpRequest(url, { responseType: 'arraybuffer' });
+			// console.log(JSON.stringify(response.data))
```
Comment (sameersubudhi) [inline validation/src/validateURLs.js:82]:
```suggestion
```

File: validation/src/validateURLs.js [hunk 4]
```diff
@@ -62,29 +64,103 @@ const validateProjectPageURL = async (projectPageURL) => {
...
+			imageBuffer = Buffer.from(response.data);
+		} else if (filePath) {
+			imageBuffer = await readFile(filePath);
+		} else {
+			throw Error('Either url or filePath needs to be supplied to check image resolution');
```
Comment (sameersubudhi) [inline validation/src/validateURLs.js:87]:
```suggestion
			throw Error('Either URL or filePath needs to be supplied to verify the image resolution.');
```

File: validation/src/validateURLs.js [hunk 5]
```diff
@@ -62,29 +64,103 @@ const validateProjectPageURL = async (projectPageURL) => {
...
+		}
+
+		// Use sharp to get the image dimensions
+		const metadata = await sharp(imageBuffer).metadata();
+		const { width, height } = metadata;
+
+		if (width !== config.image.DEFAULT_HEIGHT || height !== config.image.DEFAULT_WIDTH) {
+			validationErrors.push(`Error validating logo URL: ${url}. Image resolution should be ${config.image.DEFAULT_HEIGHT}x${config.image.DEFAULT_WIDTH}.`);
```
Comment (sameersubudhi) [inline validation/src/validateURLs.js:95]:
```suggestion
			validationErrors.push(`Error validating logo URL: ${url}. Image resolution must be ${config.image.DEFAULT_HEIGHT}px x ${config.image.DEFAULT_WIDTH}px.`);
```

File: validation/src/validateURLs.js [hunk 6]
```diff
@@ -62,29 +64,103 @@ const validateProjectPageURL = async (projectPageURL) => {
...
 	const validationErrors = [];
 	const { png: pngURL, svg: svgURL } = logos;
 
 	if (pngURL) {
-		try {
-			await httpRequest(pngURL);
-		} catch (error) {
-			validationErrors.push(new Error(`Error validating logo png URL. Error: ${error.message}.`));
+		if (!pngURL.endsWith('.png')) {
+			validationErrors.push(`Error validating logo png URL: ${pngURL}. Provided URL is not in png format.`);
```
Comment (sameersubudhi) [inline validation/src/validateURLs.js:110]:
```suggestion
			validationErrors.push(`Error validating the PNG logo URL: ${pngURL}. Provided URL is not in PNG format.`);
```

File: validation/src/validateURLs.js [hunk 7]
```diff
@@ -62,29 +64,103 @@ const validateProjectPageURL = async (projectPageURL) => {
...
+						// If logo is part of main branch but pushed in current branch
+						const filePath = pngURL.replace(`${config.repositoryURL}/${config.repositoryDefaultBranch}/`, '');
+						if (!allChangedFiles.includes(filePath)) {
+							// If logo is already part of main branch
+							validationErrors.push(...await validateImageResolution({ url: pngURL }));
+						} else {
+							validationErrors.push(...await validateImageResolution({ filePath: path.join(config.rootDir, filePath) }));
+						}
+					} else { // If logo url dosent refrance default branch or commit hash
+						validationErrors.push(`Error validating logo png URL: ${pngURL}. URL needs to be associated with ${config.repositoryDefaultBranch} branch or a commit hash.`);
```
Comment (sameersubudhi) [inline validation/src/validateURLs.js:126]:
```suggestion
						validationErrors.push(`Error validating the PNG logo URL: ${pngURL}. URL needs to be associated with the '${config.repositoryDefaultBranch}' branch or a commit hash.`);
```

File: validation/src/validateURLs.js [hunk 8]
```diff
@@ -62,29 +64,103 @@ const validateProjectPageURL = async (projectPageURL) => {
...
+					}
+				} else {
+					validationErrors.push(...await validateImageResolution({ url: pngURL }));
+				}
+			} catch (error) {
+				validationErrors.push(`Error validating logo png URL: ${pngURL}. Error: ${error.message}.`);
```
Comment (sameersubudhi) [inline validation/src/validateURLs.js:132]:
```suggestion
				validationErrors.push(`Error validating the PNG logo URL: ${pngURL}. Error: ${error.message}.`);
```

File: validation/src/validateURLs.js [hunk 9]
```diff
@@ -62,29 +64,103 @@ const validateProjectPageURL = async (projectPageURL) => {
...
 		}
 	}
 
 	if (svgURL) {
-		try {
-			await httpRequest(svgURL);
-		} catch (error) {
-			validationErrors.push(new Error(`Error validating logo svg URL. Error: ${error.message}.`));
+		if (!svgURL.endsWith('.svg')) {
+			validationErrors.push(`Error validating logo svg URL: ${svgURL}. Provided URL is not in svg format.`);
```
Comment (sameersubudhi) [inline validation/src/validateURLs.js:139]:
```suggestion
			validationErrors.push(`Error validating the SVG logo URL: ${svgURL}. Provided URL is not in SVG format.`);
```

File: validation/src/validateURLs.js [hunk 10]
```diff
@@ -62,29 +64,103 @@ const validateProjectPageURL = async (projectPageURL) => {
...
+
+						const filePath = svgURL.replace(`${config.repositoryURL}/${config.repositoryDefaultBranch}/`, '');
+						if (!allChangedFiles.includes(filePath)) {
+							// If logo is already part of main branch
+							validationErrors.push(...await validateImageResolution({ url: svgURL }));
+						} else {
+							validationErrors.push(...await validateImageResolution({ filePath: path.join(config.rootDir, filePath) }));
+						}
+					} else { // If logo url dosent refrance default branch or commit hash
+						validationErrors.push(`Error validating logo svg URL: ${svgURL}. URL needs to be associated with ${config.repositoryDefaultBranch} branch or a commit hash.`);
```
Comment (sameersubudhi) [inline validation/src/validateURLs.js:156]:
```suggestion
						validationErrors.push(`Error validating the SVG logo URL: ${svgURL}. URL needs to be associated with the '${config.repositoryDefaultBranch}' branch or a commit hash.`);
```

File: validation/src/validateURLs.js [hunk 11]
```diff
@@ -62,29 +64,103 @@ const validateProjectPageURL = async (projectPageURL) => {
...
+					}
+				} else {
+					validationErrors.push(...await validateImageResolution({ url: svgURL }));
+				}
+			} catch (error) {
+				validationErrors.push(`Error validating logo svg URL: ${svgURL}. Error: ${error.message}.`);
```
Comment (sameersubudhi) [inline validation/src/validateURLs.js:162]:
```suggestion
				validationErrors.push(`Error validating the SVG logo URL: ${svgURL}. Error: ${error.message}.`);
```

**Review by sameersubudhi — CHANGES_REQUESTED** [review]

**Review by priojeetpriyom — APPROVED** [review]

**Review by sameersubudhi — APPROVED** [review]
Nice job! LGTM 🚀

**Pull request merged**